dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1092 @@
+  using Preprocessor = identity;
+  using Postprocessor = identity;
+
+  bool HasPreprocessingParams() const { return false; }
+  bool HasPostprocessingParams() const { return false; }
+
+  template <bool value>
+  using bool_const = std::integral_constant<bool, value>;
+
+  /**
+   * The functions below get the preprocessors and postprocessors.
+   * Note the is_first/is_last compile-time parameters. Overload resolution will pick the
+   * right function with distinct return type. The overloads with false_type are final -
+   * they should not be shadowed by the Actual class. They return trivial pre/postprocessors
+   * so that the implementation of intermediate stages runs faster.
+   *
+   * When redefining these functions in Actual, the Preprocessor, PreprocessorBank and
+   * Postprocessor types must be redefined to math the return types of the respective functions
+   * or memory corruption may occur.
+   */
+
+  /// Can be redefined in CRTP derived class
+  template <int non_reduced_dim>
+  PreprocessorBank<non_reduced_dim> *GetPreprocessorBanks(
+      WorkArea &wa, int reduced_axis, std::true_type is_first) const {
+    return nullptr;
+  }
+
+  /// DO NOT redefine in CRTP derived class
+  template <int non_reduced_dim>
+  IdentityPreprocessor<non_reduced_dim> *GetPreprocessorBanks(
+      WorkArea &wa, int reduced_axis, std::false_type is_first) const {
+    return nullptr;
+  }
+
+  /// Can be redefined in CRTP derived class
+  Preprocessor *GetPreprocessors(WorkArea &wa, std::true_type is_first) const {
+    return nullptr;
+  }
+
+  /// DO NOT redefine in CRTP derived class
+  identity *GetPreprocessors(WorkArea &wa, std::false_type is_first) const {
+    return nullptr;
+  }
+
+  /// Can be redefined in CRTP derived class
+  Preprocessor GetPreprocessor(std::true_type is_first) const {
+    return {};
+  }
+
+  /// DO NOT redefine in CRTP derived class
+  identity GetPreprocessor(std::false_type is_first) const {
+    return {};
+  }
+
+  /// Can be redefined in CRTP derived class
+  Postprocessor *GetPostprocessors(WorkArea &wa, std::true_type is_last) const {
+    return nullptr;
+  }
+
+  /// DO NOT redefine in CRTP derived class
+  identity *GetPostprocessors(WorkArea &wa, std::false_type is_last) const {
+    return nullptr;
+  }
+
+  /// Can be redefined in CRTP derived class
+  Postprocessor GetPostprocessor(std::true_type is_last) const {
+    return {};
+  }
+
+  /// DO NOT redefine in CRTP derived class
+  identity GetPostprocessor(std::false_type is_last) const {
+    return {};
+  }
+
+  void CalculateTempBuffers(ScratchpadEstimator &se) {
+    int nstages = stages_.size();
+    buffer_sizes_ = {};
+    for (int stage_idx = 0; stage_idx < nstages; stage_idx++) {
+      auto &stage = stages_[stage_idx];
+      TempBufferSizes stage_buffers;
+      CalculateTempBuffers(stage_buffers, stage);
+      buffer_sizes_.Reserve(stage_buffers);
+    }
+
+    // now reserve the scratchpad - overaligned to 64 bytes, just in case
```

Comment:
In case of what? It seem like a war for some access beyond the array.

Reply:
The calculation of buffer size happens elsewhere and here we don't know what types are allocated - so we're very conservative about alignment.

Reply:
So I would rephrase the comment to explain that.